crates/cluster/src/commands/mod.rs

```diff
@@ -0,0 +1,190 @@
+use {
+    derive_more::AsRef,
+    itertools::Itertools,
+    std::path::PathBuf,
+    tokio::{fs::File, io::AsyncReadExt},
+    wcn_cluster::{
+        smart_contract::{
+            self,
+            evm::{RpcProvider, SmartContract},
+            Connector,
+        },
+        EncryptionKey,
+        Node,
+        NodeOperator,
+    },
+};
+
+pub mod deploy;
+pub mod key;
+pub mod maintenance;
+pub mod migration;
+pub mod operator;
+pub mod settings;
+pub mod show;
+pub mod signer;
+
+#[allow(clippy::large_enum_variant)]
+#[derive(clap::Subcommand, Debug)]
+pub enum SubCmd {
+    Signer(signer::SignerCmd),
+    Show(show::ShowCmd),
+    Migration(migration::MigrationCmd),
+    Maintenance(maintenance::MaintenanceCmd),
+    Operator(operator::OperatorCmd),
+    Settings(settings::SettingsCmd),
+    Deploy(deploy::DeployCmd),
+    Key(key::KeyCmd),
+}
+
+#[derive(Debug, clap::Args)]
+pub struct SharedArgs {
+    /// Private key used to sign transactions.
+    #[arg(
+        env = "WCN_SIGNER_PRIVATE_KEY", 
+        short = 's', // TODO: consider document as env var using after_help 
+        long = "signer-key",
+    )]
+    signer_key: String,
+
+    #[clap(
+        short = 'p',
+        long = "provider-url",
+        env = "WCN_CLUSTER_RPC_PROVIDER_URL"
+    )]
+    provider_url: String,
+
+    #[clap(
+        short = 'a',
+        long = "contract-address",
+        env = "WCN_CLUSTER_SMART_CONTRACT_ADDRESS"
+    )]
+    contract_address: Option<smart_contract::Address>,
+
+    #[clap(
+        long = "encryption-key",
+        short = 'k',
+        env = "WCN_CLUSTER_ENCRYPTION_KEY"
+    )]
+    /// Key used to encrypt the node operator data stored on-chain.
+    encryption_key: String,
+}
+
+impl SharedArgs {
+    pub(crate) async fn provider(&self) -> anyhow::Result<RpcProvider> {
+        let signer = smart_contract::Signer::try_from_private_key(&self.signer_key)?;
+
+        let provider_url = self.provider_url.parse().unwrap();
+        let provider = RpcProvider::new(provider_url, signer).await?;
+
+        Ok(provider)
+    }
+
+    /// Create a new SmartContract client using the provided key file and
+    /// provider URL.
+    pub(crate) async fn new_client(&self) -> anyhow::Result<SmartContract> {
```

Comment:
I recommend you build `Cluster` here and return it so you can later use it everywhere. Using `SmartContract` directly won't give you pretty errors. `Cluster` makes pre-flight invariant validation for every operation.